ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java

```diff
@@ -66,7 +66,7 @@
 
 public class BlockPropagationManager {
   private static final Logger LOG = LoggerFactory.getLogger(BlockPropagationManager.class);
-
+  public static final int DEFAULT_MAX_PENDING_BLOCKS_BEFORE_RETRY = 3;
```

Comment:
Why would we wait to fall 3 blocks behind before attempting to retrieve the parent? I can see it makes sense to limit the number of parent blocks we'd try to download for a non-canonical chain before giving up finding where it connects to the existing chain but I'd have thought we'd want to be able to import new blocks as quickly as possible rather than delaying until we've fallen 3 blocks out of sync before starting requests.

Reply:
effectively DEFAULT_MAX_PENDING_BLOCKS_BEFORE_RETRY=1 ?

Reply:
So  blocks happen to arrive out of order and Besu can already handle it saving them for future import in the Pending blocks cache. When this happens, the parent seems to arrive just after it without the need of  requesting the parent to the peers. 

However, when for any reason we miss the parent, then Besu enters the saving blocks state and cannot recover. It seems to happen, for example, when we get an ommer block so that parentHash is different and we cannot process the children. 

The threshold would avoid having to request the parent again every time that a block arrives out of order, giving enough time to the parent to be imported organically. 

Also, in a conversation with Karim, he mentioned that it could potentially avoid this to be used in an eventual attack.

Reply:
I'd be requesting quite eagerly - it should be reasonably rare for blocks to arrive out of order so the additional overhead of requesting a block we wind up receiving is small, but waiting for more 3 more blocks is going to generally be over 30 seconds that we stay out of sync.

In terms of risks of attacks, I'm not sure I see it - you'd have to send me a block to get me to request a block from another peer (and besu should be requesting it from a peer that has announced it has that block which is likely only the original source for a deliberately constructed attack). So we aren't getting an amplification effect that could launch a DoS.

So yeah I'd advocate for setting the max to 1 (essentially just remove the condition entirely - if we get a block and don't have it's parent we should request the parent).

Reply:
Right, I will remove the condition entirely then.

The code is requesting the lowest pending block parent:
`pendingBlocksManager.lowestAnnouncedBlock().ifPresent(this::requestParentBlock);`

It makes sense to request the parent of the block being saved instead of the the lowest block parent. Thoughts?

Reply:
I think its reasonable to request the lowest block we need as that ensures we're requesting blocks in the order we'll need to import them (generally).  The only catch is if we're only requesting one block at a time there's a risk that the block we choose to request might not be available and we get struck repeatedly requesting it instead of moving on to one from a different fork.

Reply:
Right, I have removed the condition. The scenario you mentioned should not happen often right? Also, if that happens, the purge mechanism would remove that stuck block when the PendingBlockCache is full and  Besu would eventually request the right missing parent.

Reply:
It shouldn't happen often but that's where a potential attack could come in - I could send Besu a block with a non-existent parent and a low block number that would always be the first to be requested.  We shouldn't have to wait for the pending cache to fill up before moving on to try a different fork. I don't think that's a new problem with this PR though so could be solved separately.